optuna/distributions.py

```diff
@@ -524,11 +525,15 @@ def __init__(self, choices: Sequence[CategoricalChoiceType]) -> None:
                 warnings.warn(message)
 
         self.choices = tuple(choices)
+        self._choice_to_index = {c: i for i, c in enumerate(choices) if isinstance(c, Hashable)}
 
     def to_external_repr(self, param_value_in_internal_repr: float) -> CategoricalChoiceType:
         return self.choices[int(param_value_in_internal_repr)]
 
     def to_internal_repr(self, param_value_in_external_repr: CategoricalChoiceType) -> float:
+        if param_value_in_external_repr in self._choice_to_index:
```

Comment:
```python
import pickle


class C:
    def __init__(self, a: int):
        self.a = a


with open("foo.pkl", mode="wb") as f:
    pickle.dump(C(1), f)


class C:
    def __init__(self, a: int):
        self.a = a
        self.b = 1 / a


with open("foo.pkl", mode="rb") as f:
    c = pickle.load(f)
    assert isinstance(c, C)
    print(f"Successfully loaded with attributes {c.__dict__}")

```

According to the code above, we can keep the backward compatibility and the old version will also be regarded as instantiated from the latest `CategoricalDistribution`, but any older versions will not have `self._choice_to_index` in the attributes of `CategoricalDistribution` if users store Optuna information in the pickle format.

Shall I change the condition as follows?

```suggestion
        if hasattr(self, "_choice_to_index") and param_value_in_external_repr in self._choice_to_index:
```